internal/app/synchronizer/evaluator_client.go

```diff
@@ -134,7 +150,77 @@ func (ec *evaluatorClient) grpcEvaluate(ctx context.Context, proposals []*pb.Mat
 	return results, nil
 }
 
-func (ec *evaluatorClient) httpEvaluate(ctx context.Context, proposals []*pb.Match) (results []*pb.Match, err error) {
-	// TODO: Implement a bidirectional HTTP streaming client in v0.7
-	return nil, nil
+func (ec *evaluatorClient) httpEvaluate(ctx context.Context, proposals []*pb.Match) ([]*pb.Match, error) {
+	reqr, reqw := io.Pipe()
+	proposalIDs := getMatchIds(proposals)
+	waitc := make(chan error)
+
+	go func() {
+		var m jsonpb.Marshaler
+		defer close(waitc)
+		defer func() {
+			sendErr := reqw.Close()
+			if sendErr != nil {
+				logger.Warning("failed to close response body read closer")
+			}
+		}()
+		for _, proposal := range proposals {
+			buf, sendErr := m.MarshalToString(&pb.EvaluateRequest{Match: proposal})
+			if sendErr != nil {
+				waitc <- status.Errorf(codes.FailedPrecondition, "failed to marshal proposal to string: %s", sendErr.Error())
+				return
+			}
+			_, sendErr = io.WriteString(reqw, buf)
+			if sendErr != nil {
+				waitc <- status.Errorf(codes.FailedPrecondition, "failed to write proto string to io writer: %s", sendErr.Error())
+				return
+			}
+		}
+	}()
+
+	req, err := http.NewRequest("POST", ec.baseURL+"/v1/evaluator/matches:evaluate", reqr)
+	if err != nil {
+		return nil, status.Errorf(codes.FailedPrecondition, "failed to create evaluator http request for proposals %s: %s", proposalIDs, err.Error())
+	}
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Transfer-Encoding", "chunked")
+
+	resp, err := http.DefaultClient.Do(req.WithContext(ctx))
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to get response from evaluator for proposals %s: %s", proposalIDs, err.Error())
+	}
+	defer func() {
+		if err = resp.Body.Close(); err != nil {
+			logger.WithError(err).Warning("failed to close response body read closer")
+		}
+	}()
+
+	var got = []*pb.Match{}
+	dec := json.NewDecoder(resp.Body)
+	for {
+		var item struct {
+			Result json.RawMessage        `json:"result"`
+			Error  map[string]interface{} `json:"error"`
+		}
+		err = dec.Decode(&item)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, status.Errorf(codes.Unavailable, "failed to read response from HTTP JSON stream: %s", err.Error())
```

Comment:
Unavailable indicates retry may get success - is that the case here?

Reply:
I feel like getting an error while reading from a JSON stream is kind of retry-able (not 100% sure).